src/razor/src/dynamicFile/dynamicFileInfoHandler.ts

```diff
@@ -33,39 +37,57 @@ export class DynamicFileInfoHandler {
+                    DynamicFileInfoHandler.dynamicFileUpdatedCommand,
+                    new RazorDynamicFileChangedParams(identifier)
+                );
+            }
+        });
     }
 
     // Given Razor document URIs, returns associated generated doc URIs
     private async provideDynamicFileInfo(
         request: ProvideDynamicFileParams
     ): Promise<ProvideDynamicFileResponse | null> {
-        let virtualUri: DocumentUri | null = null;
+        this.documentManager.roslynActivated = true;
+        const vscodeUri = vscode.Uri.parse(request.razorDocument.uri, true);
+
+        // Normally we start receiving dynamic info after Razor is initialized, but if the user had a .razor file open
+        // when they started VS Code, the order is the other way around. This no-ops if Razor is already initialized.
+        await this.documentManager.ensureRazorInitialized();
+
+        const razorDocument = await this.documentManager.getDocument(vscodeUri);
         try {
-            const vscodeUri = vscode.Uri.parse(request.razorDocument.uri, true);
-            const razorDocument = await this.documentManager.getDocument(vscodeUri);
             if (razorDocument === undefined) {
                 this.logger.logWarning(
                     `Could not find Razor document ${vscodeUri.fsPath}; adding null as a placeholder in URI array.`
                 );
-            } else {
-                // Retrieve generated doc URIs for each Razor URI we are given
-                const virtualCsharpUri = UriConverter.serialize(razorDocument.csharpDocument.uri);
-                virtualUri = virtualCsharpUri;
+
+                return null;
             }
 
-            this.documentManager.roslynActivated = true;
+            const virtualCsharpUri = UriConverter.serialize(razorDocument.csharpDocument.uri);
 
-            // Normally we start receiving dynamic info after Razor is initialized, but if the user had a .razor file open
-            // when they started VS Code, the order is the other way around. This no-ops if Razor is already initialized.
-            await this.documentManager.ensureRazorInitialized();
+            if (this.documentManager.isRazorDocumentOpenInCSharpWorkspace(vscodeUri)) {
+                // Open documents have didOpen/didChange to update the csharp buffer. Razor
+                // does not send edits and instead lets vscode handle them.
+                return new ProvideDynamicFileResponse({ uri: virtualCsharpUri }, null);
+            } else {
+                // Closed documents provide edits since the last time they were requested since
+                // there is no open buffer in vscode corresponding to the csharp content.
+                const csharpDocument = razorDocument.csharpDocument as CSharpProjectedDocument;
+                const edits = csharpDocument.getAndApplyEdits();
+
+                return new ProvideDynamicFileResponse({ uri: virtualCsharpUri }, edits ?? null);
```

Comment:
Given this can send null, and open docs always do, and thinking about the above comment, I wonder if these two branches can be combined. ie, on open, clear the edits, then provide the dynamic file response with `edits ?? null`.